concurrency/MockWebServer.java

```diff
@@ -0,0 +1,230 @@
+        }
+
+        @Override
+        public void configure(HttpsParameters params) {
+            params.setNeedClientAuth(needClientAuth);
+        }
+    }
+
+    /**
+     * Sleep the specified amount of time, if the time value is not null
+     */
+    private void sleepIfNeeded(TimeValue timeValue) throws InterruptedException {
+        if (timeValue == null) {
+            return;
+        }
+
+        CountDownLatch latch = new CountDownLatch(1);
+        latches.add(latch);
+        try {
+            latch.await(timeValue.millis(), TimeUnit.MILLISECONDS);
+        } finally {
+            latches.remove(latch);
+        }
+    }
+
+    /**
+     * Creates a MockRequest from an incoming HTTP request, that can later be checked in your test assertions
+     */
+    private MockRequest createRequest(HttpExchange exchange) throws IOException {
+        MockRequest request = new MockRequest(exchange.getRequestMethod(), exchange.getRequestURI(), exchange.getRequestHeaders());
+        if (exchange.getRequestBody() != null) {
+            String body = Streams.copyToString(new InputStreamReader(exchange.getRequestBody()));
+            if (Strings.isEmpty(body) == false) {
+                request.setBody(body);
+            }
+        }
+        return request;
+    }
+
+    /**
+     * @return The hostname the server is bound to. Uses #InetSocketAddress.getHostString() to prevent reverse dns lookups
+     */
+    public String getHostName() {
+        return server.getAddress().getHostString();
+    }
+
+    /**
+     * @return The tcp port that the server is bound to
+     */
+    public int getPort() {
+        return server.getAddress().getPort();
+    }
+
+    /**
+     * Adds a response to the response queue that is used when a request comes in
+     * Note: Every response is only processed once
+     * @param response The created mock response
+     */
+    public void enqueue(MockResponse response) {
+        responses.add(response);
+    }
+
+    /**
+     * @return The requests that have been made to this mock web server
+     */
+    public List<MockRequest> requests() {
+        return requests;
+    }
+
+    /**
+     * Removes the first request in the list of requests and returns it to the caller.
+     * This can be used as a queue if you know the order of your requests deone.
+     */
+    public MockRequest takeRequest() {
+        return requests.remove(0);
+    }
+
+    /**
+     * Closes down the webserver. Also tries to stop all the currently sleeping requests first by counting down their respective
+     * latches.
+     */
+    @Override
+    public void close() {
+        logger.debug("Counting down all latches before terminating executor");
+        latches.forEach(CountDownLatch::countDown);
+
+        if (server.getExecutor() instanceof ExecutorService) {
+            try {
+                terminate((ExecutorService) server.getExecutor());
+            } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)